### PR TITLE
wallet: increase BnB upper limit

### PR DESCRIPTION
**Pull request by S3RK**
Successor of #23367

Increasing upper bound helps to find more changeless solutions.
This improves coin selection in the cases introduced in #24580.

### CODE REVIEW

File: src/wallet/coinselection.cpp [hunk 1]
```diff
@@ -62,9 +62,9 @@ struct {
 
 static const size_t TOTAL_TRIES = 100000;
 
-std::optional<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selection_target, const CAmount& cost_of_change)
+std::optional<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selection_target, const CAmount& upper_bound)
 {
-    SelectionResult result(selection_target);
+    SelectionResult result(selection_target, /* force_no_change=*/ true);
```
Comment (fanquake) [inline src/wallet/coinselection.cpp:67]:
```suggestion
    SelectionResult result(selection_target, /*force_no_change=*/true);
```
Anywhere you are adding/touching named args, please use the [style from the developer docs](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c-named-arguments).

**Comment by S3RK** [comment]
@achow101 take a look. I reworked my BnB optimization and all the tests pass now. I think the next step would be to do simulations and see how this affects coin selection at large (but locally it fixes some of the edge cases found in #24580)

File: src/wallet/coinselection.cpp (hunk 1 above)
Comment (S3RK) [reply src/wallet/coinselection.cpp]: fixed

File: src/wallet/coinselection.h [hunk 2]
```diff
@@ -254,6 +254,10 @@ struct SelectionResult
 private:
     /** Set of inputs selected by the algorithm to use in the transaction */
     std::set<COutput> m_selected_inputs;
+    /** The target the algorithm selected for. Note that this may not be equal to the recipient amount as it can include non-input fees */
+    CAmount m_target;
```
Comment (Xekyo) [inline src/wallet/coinselection.h:258]: I wish we had canonical terms to consistently disambiguate the selection target and the recipient output total, especially because it additionally sometimes includes cost of change and a minimum change.

What do you all think about something along the lines of:
- `recipients_total`: sum of recipient output values (excluding change)
- `selection_target`: the value the coin selection needs to raise to pay for the payload including change for algorithms that allow change

Do we need another term for `selection_target` without change?

File: src/wallet/spend.cpp [hunk 3]
```diff
@@ -434,9 +434,17 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
              */
             preset_inputs.Insert(out, /*ancestors=*/ 0, /*descendants=*/ 0, /*positive_only=*/ false);
         }
+        if (preset_inputs.GetSelectionAmount() < nTargetValue) return std::nullopt;
```
Comment (Xekyo) [inline src/wallet/spend.cpp:438]: Wasn't this what we introduced `SelectionResult.Merge(…)` above for?

File: src/wallet/spend.cpp [hunk 4]
```diff
@@ -572,8 +585,8 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
     if (!res) return std::nullopt;
 
     // Add preset inputs to result
-    res->AddInput(preset_inputs);
-    if (res->m_algo == SelectionAlgorithm::MANUAL) {
+    res->Merge(preselected);
```
Comment (Xekyo) [inline src/wallet/spend.cpp:589]: I have the impression that this commit makes changes to how coin selection works with preselected inputs and in addition introduces that SelectionResult returns change. While both changes look worthwhile to me, it would be easier to follow if they were separate commits if not separate PRs. Am I missing some special reason why doing all of this at once is advantageous?

**Review by murchandamus — COMMENTED** [review]
Very exciting changes in this PR. It feels like there are still some moving parts, see my comments on the various sections.

File: src/wallet/spend.cpp [hunk 5]
```diff
@@ -760,7 +760,8 @@ static bool CreateTransactionInternal(
 
     // vouts to the payees
     if (!coin_selection_params.m_subtract_fee_outputs) {
-        coin_selection_params.tx_noinputs_size = 11; // Static vsize overhead + outputs vsize. 4 nVersion, 4 nLocktime, 1 input count, 1 output count, 1 witness overhead (dummy, flag, stack size)
+        coin_selection_params.tx_noinputs_size = 10; // Static vsize overhead + outputs vsize. 4 nVersion, 4 nLocktime, 1 input count, 1 witness overhead (dummy, flag, stack size)
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:782]: For clarity. The size required to store output counter added below. So in case of <253 outputs `tx_noinputs_size` will be 11vB for both segwit and non-segwit txs. This overshoots the target a bit for non-segwit, but this is much better than selecting not enough coins.

File: src/wallet/coinselection.h [hunk 6]
```diff
@@ -254,6 +254,10 @@ struct SelectionResult
...
+    /** The algorithm used to produce this result */
+    SelectionAlgorithm m_algo;
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/coinselection.h:260]: Yes, I just made it non const to allow merging two `SelectionResult`s

File: src/wallet/spend.cpp [hunk 7]
```diff
@@ -382,9 +382,10 @@ std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAm
     // Vector of results. We will choose the best one based on waste.
     std::vector<SelectionResult> results;
 
+    const auto cost_of_extra_input = coin_selection_params.m_effective_feerate.GetFee(DUMMY_NESTED_P2WPKH_INPUT_SIZE);
     // Note that unlike KnapsackSolver, we do not include the fee for creating a change output as BnB will not create a change output.
     std::vector<OutputGroup> positive_groups = GroupOutputs(wallet, coins, coin_selection_params, eligibility_filter, true /* positive_only */);
-    if (auto bnb_result{SelectCoinsBnB(positive_groups, nTargetValue, coin_selection_params.m_cost_of_change)}) {
+    if (auto bnb_result{SelectCoinsBnB(positive_groups, nTargetValue, coin_selection_params.m_cost_of_change + cost_of_extra_input)}) {
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:388]: To fix inefficiencies identified #24580 we need to increase upper limit by the opportunity cost of spending one extra input. We don't know what type that one extra input would be.

The range of possible values is from the cost of p2tr (57.5vb) to the cost of p2pkh (148vb). The chosen value is in the middle of that range. 

- If you have only p2pkh inputs in your wallet: the new upper bound is lower than optimal, but still better than current bound as it extends the set of possible BnB solutions.
- If you have only p2tr inputs: the new bound is higher than optimal, but also better than current bound. Same reasoning applies here as for the BnB without upper limit. We produce more BnB solutions and later choose what solution is economically more efficient. But unlike BnB without upper limit we don't risk dropping ridiculous amounts to fees.

An alternative way to pick the value would be to examine available UTXOs and take an average input size or something like this. This is more complicated and I'm not sure if it makes a meaningful difference.

File: src/wallet/spend.cpp [hunk 8]
```diff
@@ -434,9 +434,17 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
              */
             preset_inputs.Insert(out, /*ancestors=*/ 0, /*descendants=*/ 0, /*positive_only=*/ false);
         }
+        if (preset_inputs.GetSelectionAmount() < nTargetValue) return std::nullopt;
+
+        if (preset_inputs.GetSelectionAmount() > nTargetValue + coin_selection_params.m_cost_of_change) {
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:440]: I don't think so, because `m_cost_of_change` is higher than dust

File: src/wallet/spend.cpp [hunk 9]
```diff
@@ -398,11 +398,8 @@ std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAm
         results.push_back(*knapsack_result);
     }
 
-    // Include change for SRD as we want to avoid making really small change if the selection just
-    // barely meets the target. Just use the lower bound change target instead of the randomly
-    // generated one, since SRD will result in a random change amount anyway; avoid making the
-    // target needlessly large.
-    const CAmount srd_target = nTargetValue + coin_selection_params.m_change_fee + CHANGE_LOWER;
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:405]: Will do. The reason is to have the correct value (without CHANGE_LOWER) for `SelectionResult.m_target`

File: src/wallet/spend.cpp [hunk 10]
```diff
@@ -27,25 +27,25 @@ using interfaces::FoundBlock;
 namespace wallet {
 static constexpr size_t OUTPUT_GROUP_MAX_ENTRIES{100};
 
-int GetTxSpendSize(const CWallet& wallet, const CWalletTx& wtx, unsigned int out, bool use_max_sig)
+int GetTxSpendSize(const CWallet& wallet, const CWalletTx& wtx, unsigned int out, const CCoinControl* coin_control)
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:30]: Will do

File: src/wallet/coinselection.h [hunk 11]
```diff
@@ -258,15 +258,17 @@ struct SelectionResult
     bool m_use_effective{false};
     /** The computed waste */
     std::optional<CAmount> m_waste;
+    /** Force no change even if more than enough inputs are selected */
+    bool m_force_no_change{false};
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/coinselection.h:266]: meh.. I'm ambivalent about this. What do others think?

File: src/wallet/coinselection.cpp [hunk 12]
```diff
@@ -439,7 +450,18 @@ void SelectionResult::Clear()
...
+void SelectionResult::Merge(const SelectionResult& other)
+{
+    m_target += other.m_target;
+    m_use_effective |= other.m_use_effective;
+    m_force_no_change |= other.m_force_no_change;
+    if (m_algo == SelectionAlgorithm::MANUAL) {
+        m_algo = other.m_algo;
+    }
+    util::insert(m_selected_inputs, other.m_selected_inputs);
 }
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/coinselection.cpp:465]: I'll work on splitting this into a separate commit

File: src/wallet/spend.cpp (hunk 3 above)
Comment (S3RK) [reply src/wallet/spend.cpp:438]: Not sure what did you mean by this. This is the case when the selected inputs are not enough and we just return null value. There's no merging going on.

File: src/wallet/spend.cpp (hunk 4 above)
Comment (S3RK) [reply src/wallet/spend.cpp:589]: I intended this to not change the behavior. The reason why we need `Merge` is to correctly track `SelectionResult::m_target`. Right now in case with preselected inputs `m_target` will be reduced by the amount of preselected inputs and this leads to incorrect result in `SelectionResult::GetChange`

I'll split the commits and try to explain better why this is needed

File: src/wallet/spend.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:388]: I should probably add a comment to the code once we agree on the value

File: src/wallet/spend.cpp (hunk 3 above)
Comment (Xekyo) [reply src/wallet/spend.cpp:438]: Right, but if the preset inputs are less than the target, wouldn't we be supposed to select additional inputs and build a transaction? I thought that's what you introduced the merge for.

File: src/wallet/spend.cpp (hunk 4 above)
Comment (Xekyo) [reply src/wallet/spend.cpp:589]: Thanks

**Comment by Xekyo** [comment]
branch name, simulation hash, payment sequence, feerate sequence, and scenario

Update:
[I did some simulations](https://github.com/achow101/coin-selection-simulation/pull/10/commits/cbe0346dcbe70dfa862da3abb13875e7ae3b1c0a) for this branch.
The results were that
 
 -  the cost increased when the BnB window was relaxed
 - the number of changeless solutions was reduced
 
 FWIU, @S3RK had a similar result with his simulations so far. There was a bug discovered in how the waste is calculated for Knapsack in some instances, but I don't expect this bug to impact the overall trend.
 
More experiments to follow, and it may be interesting to explore the opposite, i.e. see  whether reducing the window size in which BnB produces solutions improves results instead.

**Comment by S3RK** [comment]
> The results were that
> 
> * the cost increased when the BnB window was relaxed
> * the number of changeless solutions was reduced

I don't think this is the correct interpretation of the data. e.g. for `bustabit-tiny-hot-wallet_2019-2020-per-day`
* total cost remained the same.
   * average balance: baseline - 14.6401 vs branch - 14.6396
   * average total fees: baseline - 0.4950 vs branch - 0.4954
   * average total cost: baseline - 0.4958 vs branch - 0.4963
* the number of changeless solutions very slightly increased (~ +3% but we see more knapsack solutions and it's not clear if the difference is significant or due to randomness)
   * average for baseline - 1668.2 vs branch - 1721.6

I see similar picture for `bustabit-tiny-hot-wallet_peak-and-tail` and for my own simulations.

Overall it's not clear whether increasing the upper limit is beneficial. I'm closing this now and will send refactoring PRs separately.

**Comment by S3RK** [comment]
Close
